janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -103,45 +109,140 @@ public static JanusGraphTransaction getTx(Traversal.Admin<?, ?> traversal) {
+        } while (currentStep instanceof RepeatStep);
+
+        while (!rawLocations.isEmpty()) {
+            currentStep = rawLocations.poll();
+            Optional<Step> positionInLocalTraversal = getLocalMultiQueryPositionForStep(currentStep);
+            if (positionInLocalTraversal.isPresent()) {
+                multiQueryStepLocations.add(positionInLocalTraversal.get());
+            } else {
+                rawLocations.add(currentStep.getTraversal().getParent().asStep());
             }
         }
-        return new ArrayList<>(multiQueryCompatibleSteps);
+
+        return multiQueryStepLocations;
+    }
+
+    /**
+     * For a MultiQuery compatible step, this method searches the correct position in the step's traversal at which
+     * a <code>JanusGraphMultiQueryStep</code> should be inserted. Only the traversal of the given step is considered,
+     * parent and child traversals are not taken into account.
+     * @param step The MultiQuery compatible step.
+     * @return The step before which a <code>JanusGraphMultiQueryStep</code> should be inserted.
+     * @see org.janusgraph.graphdb.tinkerpop.optimize.step.JanusGraphMultiQueryStep
+     */
+    public static Optional<Step> getLocalMultiQueryPositionForStep(Step<?, ?> step) {
+        Step currentStep = step;
+        Step previousStep = step.getPreviousStep();
+        while ((previousStep instanceof SideEffectStep || previousStep instanceof ProfileStep)
+            && !(previousStep instanceof InjectStep)) {
+            currentStep = previousStep;
+            previousStep = previousStep.getPreviousStep();
+        }
+        if (previousStep.equals(EmptyStep.instance()) || previousStep instanceof StartStep) {
+            final Step parentStep = step.getTraversal().getParent().asStep();
+            if (isMultiQueryCompatibleParent(parentStep) && !(parentStep instanceof RepeatStep)) {
+                return Optional.empty(); // no position found for JanusGraphMultiQueryStep in this local traversal
+            } else {
+                return Optional.of(currentStep); // place JanusGraphMultiQueryStep at the stat of the local traversal
+            }
+        } else if (previousStep instanceof NoOpBarrierStep) {
+            return Optional.of(previousStep);
+        } else {
+            return Optional.of(currentStep);
+        }
+    }
+
+    /**
+     * Checks whether this step is a traversal parent for which a preceding <code>JanusGraphMultiQueryStep</code>
+     * can have a positive impact on the step's child traversals.
+     * @param step The step to be checked.
+     * @return <code>true</code> if the step's child traversals can possibly benefit of a preceding
+     *         <code>JanusGraphMultiQueryStep</code>, otherwise <code>false</code>.
+     */
+    public static boolean isMultiQueryCompatibleParent(Step<?, ?> step) {
+        return MULTIQUERY_COMPATIBLE_PARENTS.stream().anyMatch(stepClass -> stepClass.isInstance(step));
```

Comment:
Probably the below check will be faster and we don't create a new spliterator but on another hand your code is cleaner. Thus, I don't have strong preference. It's up to you.
```
return step instanceof BranchStep || step instanceof OptionalStep || step instanceof TraversalFilterStep || step instanceof RepeatStep;
```

Reply:
How about a simple `for` loop over `MULTIQUERY_COMPATIBLE_PARENTS`?

Reply:
Sounds good to me